Source/Core/DolphinQt2/FIFOPlayerWindow.cpp

```diff
@@ -0,0 +1,359 @@
+
+  bool result = file->Save(path.toStdString());
+
+  if (!result)
+    QMessageBox::critical(this, tr("Error"), tr("Failed to save FIFO log."));
+}
+
+void FIFOPlayerWindow::ToggleRecording()
+{
+  FifoRecorder& recorder = FifoRecorder::GetInstance();
+
+  if (!recorder.IsRecording())
+  {
+    // Start recording
+    recorder.StartRecording(m_frame_record_count->value(),
+                            [this] { QueueOnObject(this, [this] { OnRecordingDone(); }); });
+
+    m_save->setEnabled(false);
+    UpdateControls(Core::IsRunning());
+
+    m_record->setText(tr("Stop"));
+  }
+  else
+  {
+    recorder.StopRecording();
+  }
+
+  UpdateInfo();
+}
+
+void FIFOPlayerWindow::OnEmulationStarted()
+{
+  UpdateControls(true);
+}
+
+void FIFOPlayerWindow::OnEmulationStopped()
+{
+  // If we have previously been recording, stop now.
+  if (FifoRecorder::GetInstance().IsRecording())
+    ToggleRecording();
+
+  UpdateControls(false);
+}
+
+void FIFOPlayerWindow::OnRecordingDone()
+{
+  FifoDataFile* file = FifoRecorder::GetInstance().GetRecordedFile();
+  size_t fifo_bytes = 0;
+  size_t mem_bytes = 0;
+
+  for (u32 i = 0; i < file->GetFrameCount(); ++i)
+  {
+    fifo_bytes += file->GetFrame(i).fifoData.size();
+    for (const auto& mem_update : file->GetFrame(i).memoryUpdates)
+      mem_bytes += mem_update.data.size();
+  }
+
+  m_info_label->setText(tr("%1 FIFO bytes\n%2 memory bytes\n%3 frames")
+                            .arg(QString::number(fifo_bytes), QString::number(mem_bytes),
+                                 QString::number(file->GetFrameCount())));
+
+  m_save->setEnabled(true);
+  UpdateControls(Core::IsRunning());
+
+  m_record->setText(tr("Record"));
+}
+
+void FIFOPlayerWindow::UpdateInfo()
+{
+  if (FifoPlayer::GetInstance().GetFile() != nullptr && Core::IsRunning())
+  {
+    FifoDataFile* file = FifoPlayer::GetInstance().GetFile();
+    m_info_label->setText(
+        tr("%1 frame(s)\n%2 object(s)\nCurrent Frame: %3")
+            .arg(QString::number(file->GetFrameCount()),
+                 QString::number(FifoPlayer::GetInstance().GetFrameObjectCount()),
+                 QString::number(FifoPlayer::GetInstance().GetCurrentFrameNum())));
+    return;
+  }
+
+  if (FifoRecorder::GetInstance().IsRecording())
+  {
+    m_info_label->setText(tr("Recording..."));
+    return;
+  }
+
+  m_info_label->setText(tr("No file loaded / recorded."));
+}
+
+void FIFOPlayerWindow::OnFIFOLoaded()
+{
+  FifoDataFile* file = FifoPlayer::GetInstance().GetFile();
+
+  auto object_count = FifoPlayer::GetInstance().GetFrameObjectCount();
+  auto frame_count = file->GetFrameCount();
+
+  m_frame_range_to->setMaximum(frame_count);
+  m_frame_range_to->setValue(frame_count);
+  m_object_range_to->setMaximum(FifoPlayer::GetInstance().GetFrameObjectCount());
+  m_object_range_to->setValue(object_count);
+
+  UpdateInfo();
+  UpdateLimits();
+  UpdateControls(Core::IsRunning());
+}
+
+void FIFOPlayerWindow::OnFrameFromChanged(int value)
+{
```

Comment:
Each of these methods `OnFrameFromChanged`, `OnFrameToChanged`, `OnObjectFromChanged`, and `OnObjectToChanged` do basically the same thing, but they look different! Can you clean this up? Or, I'd even stick these all in the same function to so that similar functionality is in one place.

Reply:
They do similar things, but they're far from the same. I wouldn't now how to generalize that.

Reply:
I mean that they all, basically, pass the argument to a FifoPlayer method, then call `UpdateLimits()`, but they're written differently. Some assign the FifoPlayer to a reference, some don't, and there's varying amounts of whitespace. They should be consistent.

You could also write something like the following rather than having separate methods:

```c++
void FIFOPlayerWindow::OnLimitsChanged()
{
  FifoPlayer& player = FifoPlayer::GetInstance();
  player.SetFrameRangeStart(m_frame_range_from->value());
  player.SetFrameRangeEnd(m_frame_range_to->value());
  player.SetObjectRangeStart(m_object_range_from->value());
  player.SetObjectRangeEnd(m_object_range_to->value());

  UpdateLimits();
}
```